Simple conflict-resolution for JSON symbol tables

xbauch opened the pull request:
Link JSON symbol tables via `linking(old_symtab, new_symtab, message_handler)`.

Instead of trying to merge them and throwing exception for every symbol
conflict.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton commented:
You might want to look at whether the linker used for combining multiple C symbol tables suits your needs

xbauch commented:
@smowton Is this what you had in mind?

petr-bauch committed:
Link JSON symbol tables

Instead of trying to merge them and throwing exception for every symbol
conflict.

smowton commented:
If it suits your needs, yes

tautschnig reviewed (approved):
Is there a chance to have this tested in some way? Though what I'm more worried about: the linker currently applies conflict resolution rules that seem acceptable with C code. Other languages might be more strict about those conflicts.

petr-bauch committed:
Add regression test for linking JSONs

generated from Ada files. These are included as a documentation.

xbauch commented:
@tautschnig I've added a regression test here: https://github.com/diffblue/cbmc/pull/4613/commits/a88a188245388dd32e738b2ef5f72aef0175c4cf. And then found out that there is a discrepancy between the condition in `remove_returns` and its validation check. The last commit (https://github.com/diffblue/cbmc/pull/4613/commits/16e2a14009d66c60f480e4804d8201d76bc4ee60) unifies them but I can also make it into a separate PR if that would be more appropriate.

petr-bauch committed:
Unify the remove-returns query

between goto-program/remove-return and the respective validation check. Update
the unit-test accordingly.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 16e2a14009d66c60f480e4804d8201d76bc4ee60).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111724200
